File: ClipmanMac/Sources/Clipman/HistoryTabs.swift
import Foundation

enum HistoryTabID {
    static let text = "Text"
    static let links = "Links"
    static let richText = "RichText"
    static let files = "Files"
    static let defaultOrder = [text, links, richText, files]

    static func normalizeOrder(_ values: [String]?) -> [String] {
        var result: [String] = []
        for value in values ?? [] {
            guard let canonical = canonical(value), !result.contains(where: { $0.caseInsensitiveCompare(canonical) == .orderedSame }) else { continue }
            result.append(canonical)
        }
        for value in defaultOrder where !result.contains(where: { $0.caseInsensitiveCompare(value) == .orderedSame }) {
            result.append(value)
        }
        return result
    }

    static func visibleOrder(_ values: [String]?, linksEnabled: Bool, richTextEnabled: Bool) -> [String] {
        normalizeOrder(values).filter { value in
            (value != links || linksEnabled) && (value != richText || richTextEnabled)
        }
    }

    static func moving(_ values: [String]?, selected: String, direction: Int, linksEnabled: Bool, richTextEnabled: Bool) -> [String]? {
        var result = normalizeOrder(values)
        let visible = visibleOrder(result, linksEnabled: linksEnabled, richTextEnabled: richTextEnabled)
        guard direction != 0,
              let selectedIndex = visible.firstIndex(where: { $0.caseInsensitiveCompare(selected) == .orderedSame })
        else { return nil }
        let targetIndex = selectedIndex + (direction < 0 ? -1 : 1)
        guard visible.indices.contains(targetIndex),
              let first = result.firstIndex(of: visible[selectedIndex]),
              let second = result.firstIndex(of: visible[targetIndex])
        else { return nil }
        result.swapAt(first, second)
        return result
    }

    static func normalize(_ value: String?, linksEnabled: Bool, richTextEnabled: Bool = false) -> String {
        let trimmed = (value ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.caseInsensitiveCompare(files) == .orderedSame {
            return files
        }
        if linksEnabled, trimmed.caseInsensitiveCompare(links) == .orderedSame {
            return links
        }
        if richTextEnabled, trimmed.caseInsensitiveCompare(richText) == .orderedSame {
            return richText
        }
        return text
    }

    private static func canonical(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return defaultOrder.first { $0.caseInsensitiveCompare(trimmed) == .orderedSame }
    }
}

File: src/HistoryTabs.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Clipman
{
    internal static class HistoryTabs
    {
        public const string Text = "Text";
        public const string Links = "Links";
        public const string RichText = "RichText";
        public const string Files = "Files";

        public static List<string> DefaultOrder()
        {
            return new List<string> { Text, Links, RichText, Files };
        }

        public static List<string> NormalizeOrder(IEnumerable<string> values)
        {
            var normalized = new List<string>();
            foreach (var value in values ?? Enumerable.Empty<string>())
            {
                var canonical = Canonical(value);
                if (canonical != null && !normalized.Contains(canonical, StringComparer.OrdinalIgnoreCase))
                {
                    normalized.Add(canonical);
                }
            }
            foreach (var tab in DefaultOrder())
            {
                if (!normalized.Contains(tab, StringComparer.OrdinalIgnoreCase)) normalized.Add(tab);
            }
            return normalized;
        }

        public static List<string> VisibleOrder(IEnumerable<string> values, bool linksEnabled, bool richTextEnabled)
        {
            return NormalizeOrder(values).Where(tab =>
                !string.Equals(tab, Links, StringComparison.OrdinalIgnoreCase) || linksEnabled)
                .Where(tab => !string.Equals(tab, RichText, StringComparison.OrdinalIgnoreCase) || richTextEnabled)
                .ToList();
        }

        public static bool TryMoveVisible(IEnumerable<string> values, string selected, int direction, bool linksEnabled, bool richTextEnabled, out List<string> moved)
        {
            moved = NormalizeOrder(values);
            var visible = VisibleOrder(moved, linksEnabled, richTextEnabled);
            var selectedIndex = visible.FindIndex(tab => string.Equals(tab, selected, StringComparison.OrdinalIgnoreCase));
            var targetIndex = selectedIndex + Math.Sign(direction);
            if (selectedIndex < 0 || direction == 0 || targetIndex < 0 || targetIndex >= visible.Count) return false;
            var first = moved.FindIndex(tab => string.Equals(tab, visible[selectedIndex], StringComparison.OrdinalIgnoreCase));
            var second = moved.FindIndex(tab => string.Equals(tab, visible[targetIndex], StringComparison.OrdinalIgnoreCase));
            var temporary = moved[first];
            moved[first] = moved[second];
            moved[second] = temporary;
            return true;
        }

        public static string Normalize(string value, bool linksEnabled, bool richTextEnabled)
        {
            var trimmed = (value ?? string.Empty).Trim();
            if (string.Equals(trimmed, Files, StringComparison.OrdinalIgnoreCase)) return Files;
            if (linksEnabled && string.Equals(trimmed, Links, StringComparison.OrdinalIgnoreCase)) return Links;
            if (richTextEnabled && string.Equals(trimmed, RichText, StringComparison.OrdinalIgnoreCase)) return RichText;
            return Text;
        }

        private static string Canonical(string value)
        {
            var trimmed = (value ?? string.Empty).Trim();
            if (string.Equals(trimmed, Text, StringComparison.OrdinalIgnoreCase)) return Text;
            if (string.Equals(trimmed, Links, StringComparison.OrdinalIgnoreCase)) return Links;
            if (string.Equals(trimmed, RichText, StringComparison.OrdinalIgnoreCase)) return RichText;
            if (string.Equals(trimmed, Files, StringComparison.OrdinalIgnoreCase)) return Files;
            return null;
        }
    }
}
